source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -211,24 +212,25 @@ FilterStatus Router::messageBegin(MessageMetadataSharedPtr metadata) {
 
   route_entry_ = route_->routeEntry();
 
-  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(route_entry_->clusterName());
+  const std::string& cluster_name = getClusterName(metadata);
```

Comment:
The cluster lookup is also performed in the ratelimit filter (and potentially custom filters).

The intention of the design is that `RouteEntry::clusterName()` returns the selected cluster name no matter how it was chosen. So I think the correct way to implement this is to introduce a new class, similar to `WeightedClusterEntry`, constructed with a parent `RouteEntryImplBase*` and a cluster name. (This all works as in the HTTP router where the analogous class is called `DynamicRouteEntry`). 

In `RouteEntryImplBase::clusterEntry` you can then detect the cluster_header config and extract the chosen cluster name and return a newly constructed `DynamicRouteEntry` (you'll have to plumb through the metadata, but I think it's available at all the clusterEntry call sites).

Then this line revert to calling `route_entry_->clusterName()` and any other calls to `RouteEntry::clusterName()` will get the correct cluster name. (It's fine to cache the name in this method for logging.)

Reply:
Ah -- I see. Thanks for the quick review and the detailed explanation, I'll rework this to follow that model.